Client/Client.Core/Client.Core.Static/Client.cpp

```diff
@@ -657,4 +706,92 @@ void Client::onDirectMessageCreateAnswer(Utility::DirectMessageStatus directMess
         Base::Logger::LogLevel::WARNING
     );
 }
+
+void Client::onConnectionInfoAnswer(const Models::ConnectionInfo& connectionInfo)
+{
+    _connectionInfo = connectionInfo;
+
+    _connection->setConnectionVerifier(std::make_shared<Base::Verifiers::HashVerifier>());
+}
+
+void Client::onKeyAgreement(const Models::KeyAgreementInfo& serverKeyAgreementInfo)
+{
+    bool isSharedSecretCalculated = false;
+    CryptoPP::SecByteBlock sharedSecret;
+
+    if (!serverKeyAgreementInfo._publicKey.empty())
+    {
+        std::string publicServerKeyStr = serverKeyAgreementInfo._publicKey;
+        CryptoPP::SecByteBlock publicServerKey(reinterpret_cast<const CryptoPP::byte*>(publicServerKeyStr.data()), publicServerKeyStr.size());
+        sharedSecret = _connection->getKeyAgreement()->calculateSharedKey(publicServerKey);
+
+        isSharedSecretCalculated = !sharedSecret.empty();
+    }
+    else
+    {
+        Base::FileLogger::getInstance().log("Empty public server key. ", Base::LogLevel::WARNING);
+    }
+
+    if (isSharedSecretCalculated && !serverKeyAgreementInfo._publicKey.empty() &&
+        Base::SessionKeyHolder::Instance().setKey(std::move(sharedSecret), _connection->getUserID()) == Utility::GeneralCodes::SUCCESS)
+    {
+        _connection->setEncryption(std::make_shared<Base::Crypto::Symmetric::AES_GCM>());
+        _connection->setKeyConfirmator(std::make_shared<Base::KeyConfirmators::KeyConfirmation<> >());
+
+        _connection->getKeyAgreement().~shared_ptr();
```

Comment:
Why do you do it there?

Reply:
Shared_ptr contains object, which stores template key, based on which we get encryption key. After successful encryption key generation no needs to contains them further and it isn't good to contain key, which helps to calculate encryption key.